qtgql/slot.py

```diff
@@ -1,52 +1,50 @@
 import itertools
-from typing import Any, Optional, get_args
+from typing import Any
 
 from PySide6 import QtCore as qtc
 
-from qtgql.typingref import is_optional
+from qtgql.typingref import TypeHinter
 
 
-def get_optional_args(annotations: list[Any]) -> list[Any]:
-    return [get_args(arg)[0] for arg in annotations if is_optional(arg)]
+def get_optional_args(annotations: list[TypeHinter]) -> list[TypeHinter]:
+    return [arg.of_type[0] for arg in annotations if arg.is_optional()]
 
 
-def get_concrete(annotation: Any) -> Optional[type]:
-    return getattr(annotation, "__origin__", None)
-
-
-def get_concretes(annotations: list) -> list:
+def get_concretes(annotations: list[TypeHinter]) -> list[TypeHinter]:
     ret = []
     for item in annotations:
-        if is_optional(item):
-            annotations.remove(item)
+        annotations.remove(item)
+        if item.is_optional():
             continue
-        if concrete := get_concrete(item):
-            ret.append(concrete)
-            annotations.remove(item)
+        ret.append(item)
     return ret + annotations
 
 
+def get_combos(stripped_optionals: list[TypeHinter], concretes: list[TypeHinter]):
+    required_annotations = tuple([th.type for th in concretes])
```

Comment:
```suggestion
    required_annotations = tuple([th.as_annotation() for th in concretes])
```

Reply:
pytest will fail if changed as suggested.
ImportError while loading conftest '/home/xiang/qtgql/tests/test_itemsystem/conftest.py'.
test_itemsystem/conftest.py:6: in <module>
    from qtgql.itemsystem import GenericModel, get_base_type, role
../qtgql/itemsystem/__init__.py:1: in <module>
    from .core import GenericModel, RoleDoesNotExist, asdict, get_base_type, role
../qtgql/itemsystem/core.py:164: in <module>
    class GenericModel(Generic[_TBaseType], QAbstractListModel):
../qtgql/itemsystem/core.py:269: in GenericModel
    def pop(self, index: Optional[int] = None) -> None:
../qtgql/slot.py:36: in slot
    return wrapper(func)
../qtgql/slot.py:34: in wrapper
    return qtc.Slot(*[th.as_annotation() for th in concretes], result=return_)(func)
../qtgql/slot.py:34: in <listcomp>
    return qtc.Slot(*[th.as_annotation() for th in concretes], result=return_)(func)
../qtgql/typingref.py:66: in as_annotation
    assert object_map, "can't evaluate forward refs without object_map."
E   AssertionError: can't evaluate forward refs without object_map.

Reply:
make sure you are not using `from __future__ import annotations` This should be a standalone test.

Reply:
also please push you commit here

Reply:
yes, I run `pytest test_slot.py`, it failed.

Reply:
The code in `main` branch use `getattr(annotation, "__origin__", None)`, so I think there is no need to use `as_annotation()`